expr/func_get.go

```diff
@@ -20,5 +24,32 @@ func (s FuncGet) Context(context Context) Context {
 }
 
 func (s FuncGet) Exec(cache map[Req][]models.Series) ([]models.Series, error) {
-	return cache[s.req], nil
+	series := cache[s.req]
+
+	if len(series) == 0 {
+		return series, nil
+	}
+
+	var out []models.Series
+
+	for _, serie := range series {
+		tagSplits := strings.Split(serie.Target, ";")
+
+		if serie.Tags == nil {
+			serie.Tags = make(map[string]string, len(tagSplits))
+		}
+
+		for _, tagPair := range tagSplits[1:] {
+			parts := strings.SplitN(tagPair, "=", 2)
+			if len(parts) != 2 {
+				// Shouldn't happen
+				continue
+			}
+			serie.Tags[parts[0]] = parts[1]
+		}
+		serie.Tags["name"] = tagSplits[0]
+		out = append(out, serie)
+	}
+
```

Comment:
why put all this logic in this function? seems like the series should already be in this shape when they are stored in the cache, before processing functions are run.

Reply:
It seemed like the least intrusive place to put it. It could be done after the data fetch in graphite.go::executePlan, but I already added more explicit handling logic there than I liked. I could add a function similar to `dataprocessor.go::mergeSeries` that does the tag extraction.

What do you think?

Reply:
I think this should go before the call to plan.Run, not in the get function.
the expr functions tasks are to pass through series from the input to the output, potentially performing operations on them as the data passes through.  not to set up initial state for the series.
correct me if i'm wrong, but this looks like initial state for the series that should be set up once

BTW, sidenode in executePlan, the call to mergeSeries, do we need to do anything to make it work with tags? the background of that function is when you switch the interval setting of a series, you end up with two different metricdefs for the same series, so they are merged. but obviously we shouldn't merge all series together that have the same name but different tags.

Reply:
Not sure. Many of the expr functions modify the Tags, so I don't think it's out of place for the `get` function to create them in the first place. Of course, it's definitely doable to just move it before `plan.Run`

Reply:
As far as `mergeSeries`, the target is the `NameWithTags()`, so it wouldn't over merge.

Reply:
> As far as mergeSeries, the target is the NameWithTags(), so it wouldn't over merge.

perfect :)

> Not sure. Many of the expr functions modify the Tags, so I don't think it's out of place for the get function to create them in the first place. Of course, it's definitely doable to just move it before plan.Run

keep in mind that FuncGet is not always the "source of data". for certain unit tests it's FuncMock in func_mock_test.go (grep the tests for NewMock). should this function then also set the initial tags? what if the source is another function? I think there's a difference between modifying the tag set like as in FuncAggregate.Exec and setting the initial tags.

Reply:
FuncMock also bypasses the cache, so setting the proper value in the cache won't make a difference. In that case, the tags would need to be passed in. 

We could add a function to something like dataprocessor.go (`extractTags` or something) and call that from either:
1. `server::executePlan` and `FuncMock::Exec`
2. `FuncGet::Exec` and `FuncMock::Exec`
3.  `server::executePlan` or `FuncGet::Exec` and leave the callers of `FuncMock::Exec` to specify test tags explicitly.

Reply:
> FuncMock also bypasses the cache, so setting the proper value in the cache won't make a difference. In that case, the tags would need to be passed in.

cache is nothing more than "all the series that have been loaded and are ready to go into the processing chain". same thing for FuncMock's data attribute: a series with some data ready to go into the processing chain. the implementation is a bit different because of different situation (live requests in running process vs unit tests) but conceptually, very similar.

how about we add it as a method in api/models/series.go, like `Series.SetTags()`
I'm OK with 2 as long as you add a note to the Tags Field on the Series model that the property must be set by calling the SetTags method and then it should become obvious for the reader, if they search for SetTags they can see where it is set.

> leave the callers of FuncMock::Exec to specify test tags explicitly.
i also thought about this, but i don't think we should add this concern into the tests for processing functions. they should be able to assume tags are correctly set.